### ISSUE TITLE
Large Inference Time Discrepancy when using group convs for batching

### ISSUE THREAD

**Issue by mostafarohani** [pytorch/pytorch#35276]
## 🐛 Bug

Rather than using F.conv2d to convolve a (B, C, H, W) with a (C, C, K, K) kernel, imagine we reshape the input to (1, B*C, H, W) and tile the kernel to (B * C, C, K, K). We then can call F.conv2d with groups=B, and then reshape the output to be once again (B, C, H, W). These two different ways of implementing the batch convolution are the same in that they yield the same output and are doing the same computation, yet I am seeing wide fluctuations in inference depending on choices for B, C, and H/W.

For example for B = 16; C = 32; W = H = 512, the group conv approach is ~13% faster
But for say B = 64; C = 64, W = H = 64; the group conv approach is ~32% slower

## To Reproduce

1. Run the below code segment
```
import torch
import torch.nn.functional as F
import time

def normal_conv(x, weight, bias):
    return F.conv2d(x, weight, bias, stride=1, padding=1, dilation=1, groups=1)
def group_conv_hack(x, weight, bias):
    B, C, H, W = x.shape
    x = x.view(1, B * C, H, W)
    weight = weight.repeat(B, 1, 1, 1)
    bias = bias.repeat(B)
    out = F.conv2d(x, weight, bias, stride=1, padding=1, dilation=1, groups=B)
    return out.permute([1, 0, 2, 3]).view(B, C, out.shape[-2], out.shape[-1])

# modulate the below variables to observe different timing behaviors
B = 64
C = 64
W = H = 64
x = torch.randn(B, C, W, H).cuda()

num_iters = 100
weights = [torch.randn(C, C, 3, 3).cuda() for _ in range(num_iters)]
biases = [torch.randn(C).cuda() for _ in range(num_iters)]

torch.cuda.synchronize()
s = time.time()
for i in range(100):
    normal_conv(x, weights[i], biases[i])
torch.cuda.synchronize()
normal_time = time.time() - s

torch.cuda.synchronize()
s = time.time()
for i in range(100):
    group_conv_hack(x, weights[i], biases[i])
torch.cuda.synchronize()
hack_time =  time.time() - s

print(f"Normal Conv time: {normal_time} seconds")
print(f"Hack Conv time: {hack_time} seconds")
print(f"{(normal_time - hack_time)/normal_time}% faster")
```
## Expected behavior

The code snippet will compare the two implementations, running each 100 times, and compare the timing.

## Environment
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Quadro RTX 6000
Nvidia driver version: 430.50
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.4.3
[pip] numpy==1.18.1
[pip] numpy-quaternion==2020.3.5.1.6.20
[pip] numpy-stl==2.10.1
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] pytorch                   1.4.0           py3.7_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] torch                     1.1.0                    pypi_0    pypi
[conda] torchvision               0.5.0                py37_cu100    pytorch

**Comment by ptrblck** [comment]
Thanks for reporting the issue.
I can reproduce the slowdown with the default setup.
However, if you use `torch.backends.cudnn.benchmark = True` and add some warmup iterations to the script, I can see a speedup of:
```python
Normal Conv time: 0.09729194641113281 seconds
Hack Conv time: 0.15732598304748535 seconds
-0.6170504224744653% faster
```
on a V100 with PyTorch build from source using CUDA10.2.89 with cudnn7.6.5.32.

**Comment by albanD** [comment]
Optimistically closing this as "cudnn default algorithm is not as good but using benchmark for it to discover a better algorithm fixes the issue".

Feel free to re-open if you think there is something actionable on pytorch side.
